### PR TITLE
fix(oracle): handle advisories contain ksplice versions

### LINKED ISSUE #1205: False ksplice vulnernerabilities reported
## Description

Scanning the latest `oraclelinux:8` image reports a number of vulnerabilities:
```
$ trivy -d i oraclelinux:8
2021-09-01T08:37:25.994-0400	DEBUG	Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2021-09-01T08:37:26.004-0400	DEBUG	cache dir:  /home/dev/.cache/trivy
2021-09-01T08:37:26.004-0400	DEBUG	DB update was skipped because DB is the latest
2021-09-01T08:37:26.004-0400	DEBUG	DB Schema: 1, Type: 1, UpdatedAt: 2021-09-01 08:54:41.313293891 +0000 UTC, NextUpdate: 2021-09-01 14:54:41.313293691 +0000 UTC, DownloadedAt: 2021-09-01 11:03:07.172661178 +0000 UTC
2021-09-01T08:37:26.004-0400	DEBUG	Vulnerability type:  [os library]
2021-09-01T08:37:26.008-0400	DEBUG	Image ID: sha256:fcf3cbfc22ac02e416b686870463a6d0611ffa8f6ec9d24a04121d9c204e0c0a
2021-09-01T08:37:26.008-0400	DEBUG	Diff IDs: [sha256:89ca13798c53c8b13fdcb56109bc25677e22b99d4137c8a7af99d7583c1ecc28]
2021-09-01T08:37:26.009-0400	INFO	Detected OS: oracle
2021-09-01T08:37:26.009-0400	INFO	Detecting Oracle Linux vulnerabilities...
2021-09-01T08:37:26.009-0400	DEBUG	Oracle Linux: os version: 8
2021-09-01T08:37:26.009-0400	DEBUG	Oracle Linux: the number of packages: 190
2021-09-01T08:37:26.012-0400	INFO	Number of language-specific files: 0


oraclelinux:8 (oracle 8.4)
==========================
Total: 15 (UNKNOWN: 0, LOW: 0, MEDIUM: 9, HIGH: 3, CRITICAL: 3)

+-------------------+------------------+----------+-------------------+-----------------------------+---------------------------------------+
|      LIBRARY      | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |        FIXED VERSION        |                 TITLE                 |
+-------------------+------------------+----------+-------------------+-----------------------------+---------------------------------------+
| glibc             | CVE-2019-9169    | CRITICAL | 2.28-151.0.1.el8  | 2:2.28-151.0.1.ksplice2.el8 | glibc: regular-expression             |
|                   |                  |          |                   |                             | match via proceed_next_node           |
|                   |                  |          |                   |                             | in posix/regexec.c leads to           |
|                   |                  |          |                   |                             | heap-based buffer over-read...        |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-9169  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2021-3326    | HIGH     |                   |                             | glibc: Assertion failure in           |
|                   |                  |          |                   |                             | ISO-2022-JP-3 gconv module            |
|                   |                  |          |                   |                             | related to combining characters       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2021-3326  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2016-10228   | MEDIUM   |                   |                             | glibc: iconv program can hang         |
|                   |                  |          |                   |                             | when invoked with the -c option       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2016-10228 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2019-25013   |          |                   |                             | glibc: buffer over-read in            |
|                   |                  |          |                   |                             | iconv when processing invalid         |
|                   |                  |          |                   |                             | multi-byte input sequences in...      |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-25013 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2020-27618   |          |                   |                             | glibc: iconv when processing          |
|                   |                  |          |                   |                             | invalid multi-byte input              |
|                   |                  |          |                   |                             | sequences fails to advance the...     |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2020-27618 |
+-------------------+------------------+----------+                   +                             +---------------------------------------+
| glibc-common      | CVE-2019-9169    | CRITICAL |                   |                             | glibc: regular-expression             |
|                   |                  |          |                   |                             | match via proceed_next_node           |
|                   |                  |          |                   |                             | in posix/regexec.c leads to           |
|                   |                  |          |                   |                             | heap-based buffer over-read...        |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-9169  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2021-3326    | HIGH     |                   |                             | glibc: Assertion failure in           |
|                   |                  |          |                   |                             | ISO-2022-JP-3 gconv module            |
|                   |                  |          |                   |                             | related to combining characters       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2021-3326  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2016-10228   | MEDIUM   |                   |                             | glibc: iconv program can hang         |
|                   |                  |          |                   |                             | when invoked with the -c option       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2016-10228 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2019-25013   |          |                   |                             | glibc: buffer over-read in            |
|                   |                  |          |                   |                             | iconv when processing invalid         |
|                   |                  |          |                   |                             | multi-byte input sequences in...      |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-25013 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2020-27618   |          |                   |                             | glibc: iconv when processing          |
|                   |                  |          |                   |                             | invalid multi-byte input              |
|                   |                  |          |                   |                             | sequences fails to advance the...     |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2020-27618 |
+-------------------+------------------+----------+                   +                             +---------------------------------------+
| glibc-langpack-en | CVE-2019-9169    | CRITICAL |                   |                             | glibc: regular-expression             |
|                   |                  |          |                   |                             | match via proceed_next_node           |
|                   |                  |          |                   |                             | in posix/regexec.c leads to           |
|                   |                  |          |                   |                             | heap-based buffer over-read...        |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-9169  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2021-3326    | HIGH     |                   |                             | glibc: Assertion failure in           |
|                   |                  |          |                   |                             | ISO-2022-JP-3 gconv module            |
|                   |                  |          |                   |                             | related to combining characters       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2021-3326  |
+                   +------------------+----------+                   +                             +---------------------------------------+
|                   | CVE-2016-10228   | MEDIUM   |                   |                             | glibc: iconv program can hang         |
|                   |                  |          |                   |                             | when invoked with the -c option       |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2016-10228 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2019-25013   |          |                   |                             | glibc: buffer over-read in            |
|                   |                  |          |                   |                             | iconv when processing invalid         |
|                   |                  |          |                   |                             | multi-byte input sequences in...      |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2019-25013 |
+                   +------------------+          +                   +                             +---------------------------------------+
|                   | CVE-2020-27618   |          |                   |                             | glibc: iconv when processing          |
|                   |                  |          |                   |                             | invalid multi-byte input              |
|                   |                  |          |                   |                             | sequences fails to advance the...     |
|                   |                  |          |                   |                             | -->avd.aquasec.com/nvd/cve-2020-27618 |
+-------------------+------------------+----------+-------------------+-----------------------------+---------------------------------------+
```

All have a reported fix version of `2:2.28-151.0.1.ksplice2.el8`.

CVE-2019-9169 has [3 separate errata reports](https://linux.oracle.com/cve/CVE-2019-9169.html) from Oracle:
- [ELSA-2021-1585](https://linux.oracle.com/errata/ELSA-2021-1585.html)
  - This has a fix version for glibc of `glibc-2.28-151.0.1.el8`, which is the version installed
- [ELSA-2021-9280](https://linux.oracle.com/errata/ELSA-2021-9280.html)
  - This has a fix version containing `ksplice1`.  I believe it's ignored thanks to #736 .
- [ELSA-2021-9344](https://linux.oracle.com/errata/ELSA-2021-9344.html)
  - This has a fix version containing `ksplice2` which is what's being reported by trivy.

I do not know anything about ksplice, but my guess is this ksplice2 packaging should be ignored the same as ksplice1.

## What did you expect to happen?
No vulnerabilities to be reported

## What happened instead?
Vulnerabilities reported that were fixed according to 1 errata, with a ksplice2 fix version.

## Output of run with `-debug`:

See above

## Output of `trivy -v`:

```
Version: 0.19.2
Vulnerability DB:
  Type: Light
  Version: 1
  UpdatedAt: 2021-09-01 08:54:41.313293891 +0000 UTC
  NextUpdate: 2021-09-01 14:54:41.313293691 +0000 UTC
  DownloadedAt: 2021-09-01 11:03:07.172661178 +0000 UTC
```

## Additional details (base image name, container registry info...):
N/A

### PR DESCRIPTION
**Pull request by afdesk**
Improve a handling of advisories contain ksplice versions:
* when one of them doesn't have ksplice, we'll also skip it
* extract kspliceX and compare it with kspliceY in advisories
* if kspliceX and kspliceY are different, we will skip the advisory.

Fixes #1205

### CODE REVIEW

File: pkg/detector/ospkg/oracle/oracle.go [hunk 1]
```diff
@@ -44,6 +44,18 @@ func NewScanner() *Scanner {
...
+	if len(subs) != 2 {
+		return ""
+	}
+	ns := strings.Split(subs[1], ".")
+	if len(ns) == 0 {
+		return ""
+	}
+	return ns[0]
+}
+
```
Comment (knqyf263) [inline pkg/detector/ospkg/oracle/oracle.go:58]: If we can assume ksplice is separated by dot, how about this?

```suggestion
func extractKspliceNumber(v string) string {
	subs := strings.Split(strings.ToLower(v), ".")
	for _, s := range subs {
		if strings.HasPrefix(v, "ksplice") {
			return s
		}
	}
	return ""
}

```

File: pkg/detector/ospkg/oracle/oracle.go [hunk 2]
```diff
@@ -65,9 +77,17 @@ func (s *Scanner) Detect(osVer string, pkgs []ftypes.Package) ([]types.DetectedV
...
+			// extract kspliceX and compare it with kspliceY in advisories
+			// if kspliceX and kspliceY are different, we will skip the advisory
+			if isKSlice := strings.Contains(adv.FixedVersion, ".ksplice"); isKSlice != strings.Contains(pkg.Release, ".ksplice") {
 				continue
+			} else if isKSlice {
+				releaseV := extractKspliceNumber(adv.FixedVersion)
+				fixedV := extractKspliceNumber(pkg.Release)
+				if releaseV != fixedV {
+					continue
+				}
```
Comment (knqyf263) [inline pkg/detector/ospkg/oracle/oracle.go:90]: I think we can simplify it.

```suggestion
			if extractKspliceNumber(adv.FixedVersion) != extractKspliceNumber(pkg.Release) {
				continue
			}
```

File: pkg/detector/ospkg/oracle/oracle_test.go [hunk 3]
```diff
@@ -151,6 +151,95 @@ func TestScanner_Detect(t *testing.T) {
...
+				},
+			},
+			want: nil,
+		},
+		{
+			name:     "the installed version has ksplice1",
+			fixtures: []string{"testdata/fixtures/oracle8.yaml"},
+			args: args{
+				osVer: "8",
+				pkgs: []ftypes.Package{
```
Comment (knqyf263) [inline pkg/detector/ospkg/oracle/oracle_test.go:199]: We already have the same test for ksplice1. These tests are duplicate. We should add only 1 test where the installed version has ksplice2.

File: pkg/detector/ospkg/oracle/oracle_test.go [hunk 4]
```diff
@@ -219,3 +308,22 @@ func TestScanner_Detect(t *testing.T) {
...
+		{"2:2.28-151.0.1.ksplice2.el8", "2"},
+		{"2:2.28-151.0.1.ksplice2", "2"},
+		{"2:2.28-151.0.1.KSplice2", "2"},
+	}
+	for _, test := range tests {
+		if got := extractKspliceNumber(test.in); got != test.res {
+			t.Errorf("getKspliceNumber(%q) got %q, want %q", test.in, got, test.res)
+		}
+	}
+}
```
Comment (knqyf263) [inline pkg/detector/ospkg/oracle/oracle_test.go:329]: We can test this method through TestScanner_Detect

**Review by knqyf263 — CHANGES_REQUESTED** [review]
Looks good! Thanks. I suggested some ideas to simplify code.

File: pkg/detector/ospkg/oracle/oracle.go (hunk 2 above)
Comment (afdesk) [reply pkg/detector/ospkg/oracle/oracle.go:90]: sure, your was is better!

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**